share/spack/setup-env.sh

```diff
@@ -77,7 +77,7 @@ function spack {
                 _sp_arg="$1"
                 shift
             fi
-            if [ "$_sp_arg" = "-h" ]; then
+            if [[ "$_sp_arg" = "-h" || "$_sp_arg" = "--help" ]]; then
```

Comment:
The double [ thing is Bash-specific. 
```suggestion
            if [ "$_sp_arg" = "-h" ] || [ "$_sp_arg" = "--help" ]; then
```

Reply:
Do you refer to the `||` within `[[ ... ]]`? We already use `[[ ... ]]` elsewhere in this file (and others).

Reply:
I'm referring to `[[`. See https://mywiki.wooledge.org/Bashism. "[[ is not defined by POSIX, originated in ksh and is also present in zsh". If we use it elsewhere, we should change that.

Reply:
It isn't supported by dash, which is the default shell on Ubuntu.

Reply:
The default shell for users on Ubuntu (and on other Debian-based distros) is bash. `/bin/sh` might be a symlink to dash, but users get `/bin/bash` by default. Also, the docs state that setup-env.sh supports only bash and zsh (see e.g. the top of this file).

Reply:
I agree, but I think we should be moving in the direction of POSIX compliance rather than away. We have people using fish and csh and other esoteric shells, so why not dash?

Reply:
In general I agree regarding POSIX compliance. However, I feel that in this case the disadvantages of POSIX outweigh the advantages:

- The file itself will need a major rewrite anyway to become POSIX compliant (the very first non-comment line in the script, `function spack {` is already not compliant).
- Having `[[`vs. `[` has substantial advantages for users (performance as I had mentioned above). The `spack` command is already quite slow, so why add more slowness?
- More "esoteric" shells such as fish or csh will most likely require a different setup-env script in any case (which was already done for csh).

Reply:
I agree with your first 2 points, but disagree with the 3rd point. The point of having a single `setup-env.sh` script instead of a separate script for bash and zsh is that all POSIX compatible shells can share it, including dash. Yes, fish and csh are not POSIX-compatible, and require separate scripts. But dash is so similar to bash and zsh that minor changes like `[[` vs `[` are all that are required to share the same script.

Reply:
Also, can you share performance metrics for `[[` vs `[`? I wasn't aware of this difference.

Reply:
The following script is almost twice as fast on my laptop with `[[` vs. `[`:

```Bash
for i in $(seq 1 100000); do
        if [[ $i -eq 1000 ]]; then echo "Found $i"; fi
done
```

`[[`: 0.486s
`[`: 0.880s

Reply:
Also, making this script compatible with *dash* (or even with *ksh*) would be a major undertaking. The main non-trivial changes that I can see are:

- Would need to remove arrays. These are heavily used on the current script and are likely to be very painful to remove.
- Would need to remove regexes. Same as above. These don't seem to be supported in *ksh*.